### DISCUSSION TITLE
Add configuration files

### DISCUSSION THREAD

**Discussion by mathisloge** [microsoft/vcpkg#19632] [Ideas]
Some ports e.g. imgui could use a header file which configures types etc. 
I would like to have some mechanism to directly inject those configuration files for a library.  

Currently I've to create a overlay port to inject those config files. But this is tedious, as I now also have to maintain these and merge the vcpkg port regularly.

It would be great if the manifests could have some field e.g. `configurationFile` which would look like: 

```json
{
    "name": "my-top-level-proj",
    "dependencies": [
        {
            "name": "imgui",
            "configurationFiles": {
                "config": "${projectSourceDir}/imgui/config.h
            }
        }
    ]
}
```

and the port vcpkg.json contains some field e.g. `configurationFiles` aswell
```json
{
   "name": "imgui",
   "configurationFiles": [{"name": "config", "required": false}]
}
```
the portfile could then use a similar procedure as the features field. e.g. `vcpkg_check_configuration_files`. 

These files are usually installed along with the port's include files or directly modify type definitions.

**Comment by autoantwort** [comment]
Would that not be something for an "options" feature of vcpkg?

**Reply by mathisloge** [reply]
the name `configurationFiles` is definitely up for discussion. In this case, the name was only meant to clarify my concern.

**Reply by autoantwort** [reply]
I mean something like https://github.com/edhebi/vcpkg/blob/master/docs/specifications/options.md (but not exactly this proposed behavior, but something in that direction)

**Reply by mathisloge** [reply]
thanks for the hint. Wasn't aware of it. I think the draft is definitely going in the right direction. If this feature is taken into account when refining the specification, we are on a really good path.

**Reply by Osyotr** [reply]
@autoantwort I don't know where to discuss this, but why can't we have `name[feature1,feature2,option1=val1,option2=val2,...]:triplet` instead of this ugly curly braces syntax?
Also, as mentioned in options specification we still have problem with copying an entire port in case we need to patch something. I would propose the following:
1) Create overlay port for a library that needs to be patched
2) In it's portfile put something like:
```cmake
vcpkg_inherit_port(qt5-base PATCHES ...)
```
Both this features would increase flexibility significantly.

**Reply by autoantwort** [reply]
> but why can't we have name[feature1,feature2,option1=val1,option2=val2,...]:triplet instead of this ugly curly braces syntax?

I strongly prefer your syntax. The spec was only an example for the direction of the feature. There are some more points that should be different imho. Maybe I have time next week to write my own rfc
